### PR TITLE
Wrap in Python

### PR DESCRIPTION
**Pull request by Besler**
- Implements FFI for input, but not output
- Creates a dynamic library with a C ABI
- Loads the library in python with `ctypes`

**FFI prototypes**
I pass the underlying pointer through in all cases expect for the complex array, which I pass an allocated array which is to be filled.

```rust
/// Free a pointer to `Record`
pub extern "C" fn record_destroy(record: *mut Record);

/// Create default record
pub extern "C" fn record_default() -> *mut Record;

/// Read record from file
pub extern "C" fn record_read(filename: *const c_char) -> *mut Record;

/// Get the record version
pub extern "C" fn record_get_version(record: *mut Record) -> *const c_char;

/// Set the record version
pub extern "C" fn record_set_version(record: *mut Record, version: *const c_char);

/// Get the record name
pub extern "C" fn record_get_name(record: *mut Record) -> *const c_char;

/// Set the record name
pub extern "C" fn record_set_name(record: *mut Record, name: *const c_char);

/// Get the number of comments
pub extern "C" fn record_get_number_of_comments(record: *mut Record) -> size_t;

/// Get an array of comments
pub extern "C" fn record_get_comment(record: *mut Record, idx: size_t) ->*const c_char;

/// Get the number of devices
pub extern "C" fn record_get_number_of_devices(record: *mut Record) -> size_t;

/// Get the device name
pub extern "C" fn record_get_device_name(record: *mut Record, idx: size_t) -> *const c_char;

/// Get the number of entries in a device
pub extern "C" fn record_get_device_number_of_entries(record: *mut Record, idx: size_t) -> size_t;

/// Get the entry from a device
pub extern "C" fn record_get_device_entry(record: *mut Record, device_idx: size_t, entry_idx: size_t) -> *const c_char;

/// Get independent variable name
pub extern "C" fn record_get_independent_variable_name(record: *mut Record) -> *const c_char;

/// Get independent variable format
pub extern "C" fn record_get_independent_variable_format(record: *mut Record) -> *const c_char;

/// Get independent variable length
pub extern "C" fn record_get_independent_variable_length(record: *mut Record) -> size_t;

/// Get independent variable array
pub extern "C" fn record_get_independent_variable_array(record: *mut Record) -> *const c_double;

/// Get number of data arrays
pub extern "C" fn record_get_number_of_data_arrays(record: *mut Record) -> size_t;

/// Get data array name
pub extern "C" fn record_get_data_array_name(record: *mut Record, idx: size_t) -> *const c_char;

/// Get data array format
pub extern "C" fn record_get_data_array_format(record: *mut Record, idx: size_t) -> *const c_char;

/// Get data array length
pub extern "C" fn record_get_data_array_length(record: *mut Record, idx: size_t) -> size_t ;

/// Get data array
pub extern "C" fn record_get_data_array(record: *mut Record, idx: size_t, mut real: *mut c_double, mut imag: *mut c_double);
```

**Python methods**
The Python interface is rather straight forward and works through property methods.

```python
    @property
    def version(self) -> str:
        '''Get the version string'''
        return CITI_LIB.record_get_version(self.__obj).decode("utf-8")
 ```

If I was to do this again, I would not have done it this way. I would have created a C-structure and passed that across the interface, deciding which side allocates and deallocates. It would be so much easier than having all these methods to maintain.

**Resources**
https://codeburst.io/how-to-use-rust-to-extend-python-360174ee5819
https://depth-first.com/articles/2020/08/03/wrapping-rust-types-as-python-classes/

### CODE REVIEW

**Comment by Besler** [comment]
I based most of this development off these references:
- https://depth-first.com/articles/2020/08/03/wrapping-rust-types-as-python-classes/
- https://pypi.org/project/setuptools-rust/

File: ffi/python/citi/record.py [hunk 1]
```diff
@@ -0,0 +1,296 @@
+'''Record implemented through a foreign function interface
```
Comment (justkash) [inline ffi/python/citi/record.py:1]: PR #19 adds support for error handling through the ffi that can be take advantage of on the python side.

File: src/ffi.rs [hunk 2]
```diff
@@ -0,0 +1,1775 @@
+//! Foreign Function Interface
```
Comment (justkash) [inline src/ffi.rs:1]: For your reference, PR #19 makes all the changes mentioned in the comments and it also adds a few extra functions for modifying citi::Record through the ffi.

**Review by justkash — CHANGES_REQUESTED** [review]
I think the biggest thing to do would be to add error handling support for the python side. You can rebase with the changes from PR #19.

**Pull request closed without merging**